### PR TITLE
feat: add exponential functions to the Drasi scalar function library

### LINKED ISSUE #212: Feature Request: Add e (Euler's number) and exp Scalar Functions
### Overview of feature request

We need to add exponential functions to the Drasi scalar function library. This task involves implementing Euler's number constant function (`e`) and the exponential function (`exp(x)`) in the core engine, and registering them for both Cypher and GQL.

This is a great task for new contributors to learn about parameterless functions and standard math function implementation in Drasi.

### Acceptance criteria

1. **Implement Core Logic**:
   - Create `E` and `Exp` structs in `core/src/evaluation/functions/numeric/numeric.rs`.
   - Implement the `ScalarFunction` trait for both.
   - **Behavior for `e()`**:
     - Takes 0 arguments.
     - Returns the constant value of Euler's number (approx 2.71828...) as a `Float`.
   - **Behavior for `exp(n)`**:
     - Takes 1 argument (`Integer` or `Float`).
     - Returns e raised to the power of n.
     - Handles `Null` inputs by returning `Null`.

2. **Register Functions**:
   - Register in `functions-cypher/src/lib.rs` (as `e` and `exp`).
   - Register in `functions-gql/src/lib.rs` (as `e` and `exp`).

3. **Testing**:
   - Add unit tests verifying the constant value of `e`.
   - Add unit tests for `exp` with various inputs (0, 1, negative, float, null).

### Additional context

- Look at `core/src/evaluation/functions/trigonometric/trigonometric.rs` for an example of a constant function (like `pi()`).
- Look at `core/src/evaluation/functions/numeric/numeric.rs` for standard numeric function patterns.
- Ensure proper error handling for invalid argument counts or types.

### PR DESCRIPTION
**Pull request by kaffie-1517**
### Description
This PR addresses Issue #212 by adding exponential functions to the Drasi scalar function library.

### 1. Implementation Logic
- **Location:** `core/src/evaluation/functions/numeric/exp.rs`
- **Changes:**
  - Created `E` struct to return Euler's constant (`approx 2.718`).
  - Created [Exp](cci:2://file:///Users/jasmeetsingh/drasi-core/core/src/evaluation/functions/numeric/exp.rs:45:0-45:17) struct to calculate `e^n`.
  - Implemented `ScalarFunction` trait for both, following the pattern in `abs.rs` and [pi.rs](cci:7://file:///Users/jasmeetsingh/drasi-core/query-ast/src/api.rs:0:0-0:0) as suggested.

### 2. Function Registration
- **Cypher:** Registered `e` and `exp` in `functions-cypher/src/lib.rs`
- **GQL:** Registered `e` and `exp` in `functions-gql/src/lib.rs`

### 3. Testing
- **Location:** `core/src/evaluation/functions/numeric/tests/exp_tests.rs`
- **Coverage:**
  - **Constant E:** Verified [e()](cci:1://file:///Users/jasmeetsingh/drasi-core/core/src/evaluation/functions/numeric/tests/exp_tests.rs:24:0-40:1) returns the correct standard `f64` constant.
  - **Standard Inputs:** Tested [exp(0)](cci:1://file:///Users/jasmeetsingh/drasi-core/core/src/evaluation/functions/numeric/tests/exp_tests.rs:42:0-97:1), [exp(1)](cci:1://file:///Users/jasmeetsingh/drasi-core/core/src/evaluation/functions/numeric/tests/exp_tests.rs:42:0-97:1), and float inputs.
  - **Negative Inputs:** Verified [exp(-1)](cci:1://file:///Users/jasmeetsingh/drasi-core/core/src/evaluation/functions/numeric/tests/exp_tests.rs:42:0-97:1) logic to return (1/e).
  - **Edge Cases:**
    - **Nulls:** [exp(null)](cci:1://file:///Users/jasmeetsingh/drasi-core/core/src/evaluation/functions/numeric/tests/exp_tests.rs:42:0-97:1) returns [null](cci:1://file:///Users/jasmeetsingh/drasi-core/core/src/evaluation/functions/numeric/tests/abs_tests.rs:83:0-101:1).
    - **Overflow:** Verified logic for large inputs (e.g., [exp(1000)](cci:1://file:///Users/jasmeetsingh/drasi-core/core/src/evaluation/functions/numeric/tests/exp_tests.rs:42:0-97:1)) to ensure it properly returns `OverflowError`.
    - **Argument Counts:** Verified strict argument validation (0 for `e` and 1 for [exp](cci:1://file:///Users/jasmeetsingh/drasi-core/core/src/evaluation/functions/numeric/tests/exp_tests.rs:42:0-97:1)).

### Verification
All unit tests passed successfully:

Screenshot 2026-02-01 at 10 30 37 AM

### Linked Issue
Closes https://github.com/drasi-project/drasi-core/issues/212

### CODE REVIEW

**Review by amansinghoriginal — COMMENTED** [review]
.

File: core/src/evaluation/functions/numeric/exp.rs [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+        args: Vec<VariableValue>,
+    ) -> Result<VariableValue, FunctionError> {
+        if !args.is_empty() {
+            return Err(FunctionError {
+                function_name: expression.name.to_string(),
+                error: FunctionEvaluationError::InvalidArgumentCount,
+            });
+        }
+        
+        Ok(VariableValue::Float(Float::from_f64(std::f64::consts::E).unwrap()))
```
Comment (amansinghoriginal) [inline core/src/evaluation/functions/numeric/exp.rs]: This will fail CI - the project enforces `clippy::unwrap_used`. Use the `From` trait instead, which is infallible for constants: (See file pi.rs)

```rust
Ok(VariableValue::Float(Float::from(std::f64::consts::E)))
```

File: core/src/evaluation/functions/numeric/exp.rs [hunk 2]
```diff
@@ -0,0 +1,118 @@
+// Copyright 2024 The Drasi Authors.
```
Comment (amansinghoriginal) [inline core/src/evaluation/functions/numeric/exp.rs:1]: Nice work on the implementation! One structural note: this codebase uses one struct per file (see `abs.rs`, `ceil.rs`, `pi.rs`). Please split `E` into `e.rs` and keep `Exp` here.

Then in `mod.rs`:
```rust
mod e;
mod exp;
pub use e::E;
pub use exp::Exp;
```

File: core/src/evaluation/functions/numeric/exp.rs [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+                error: FunctionEvaluationError::InvalidArgumentCount,
+            });
+        }
+        match &args[0] {
+            VariableValue::Null => Ok(VariableValue::Null),
+            VariableValue::Integer(n) => {
+                 match n.as_i64() {
+                    Some(i) => {
+                        let result = (i as f64).exp();
+                        if result.is_infinite() || result.is_nan() {
```
Comment (amansinghoriginal) [inline core/src/evaluation/functions/numeric/exp.rs]: This check is redundant - `Float::from_f64()` already returns `None` for infinite/NaN values. You can remove lines 68-72 and let the `match` below handle it. See `abs.rs:51-68`.

What do you thihnk?

File: core/src/evaluation/functions/numeric/exp.rs [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+                    None => Err(FunctionError {
+                        function_name: expression.name.to_string(),
+                        error: FunctionEvaluationError::OverflowError,
+                    }),
+                }
+            },
+            VariableValue::Float(n) => {
+                match n.as_f64() {
+                    Some(f) => {
+                        let result = f.exp();
```
Comment (amansinghoriginal) [inline core/src/evaluation/functions/numeric/exp.rs]: Same here - I think we can remove the redundant `is_infinite() || is_nan()` check.
What do you think?

File: core/src/evaluation/functions/numeric/tests/exp_tests.rs [hunk 5]
```diff
@@ -0,0 +1,220 @@
+// Copyright 2024 The Drasi Authors.
```
Comment (amansinghoriginal) [inline core/src/evaluation/functions/numeric/tests/exp_tests.rs:1]: Please split into `e_tests.rs` (for `E` tests) and `exp_tests.rs` (for `Exp` tests) to match the codebase pattern.

File: core/src/evaluation/functions/numeric/tests/exp_tests.rs [hunk 6]
```diff
@@ -0,0 +1,220 @@
...
+    let context =
+        ExpressionEvaluationContext::new(&binding, Arc::new(InstantQueryClock::new(0, 0)));
+    let expression = ast::FunctionExpression {
+        name: Arc::from("e"),
+        args: vec![],
+        position_in_query: 0,
+    };
+
+    let result = func.call(&context, &expression, vec![]).await.unwrap();
+    assert_eq!(
```
Comment (amansinghoriginal) [inline core/src/evaluation/functions/numeric/tests/exp_tests.rs]: Use the existing test pattern with direct comparison: (See pi_tests file)

```rust
assert!(result.is_f64());
assert_eq!(result.as_f64().unwrap(), std::f64::consts::E);
```

File: core/src/evaluation/functions/numeric/tests/exp_tests.rs [hunk 7]
```diff
@@ -0,0 +1,220 @@
...
+
+#[tokio::test]
+async fn test_e_arg_count() {
+    let func = E {};
+    let binding = QueryVariables::new();
+    let context =
+        ExpressionEvaluationContext::new(&binding, Arc::new(InstantQueryClock::new(0, 0)));
+    let expression = ast::FunctionExpression {
+        name: Arc::from("e"),
+        args: vec![],
```
Comment (amansinghoriginal) [inline core/src/evaluation/functions/numeric/tests/exp_tests.rs]: Use `assert!(matches!(...))` for error checking - it's the pattern used throughout the test suite: (See abs_tests.rs)

```rust
assert!(matches!(
    result.unwrap_err(),
    FunctionError {
        function_name: _,
        error: FunctionEvaluationError::InvalidArgumentCount
    }
));
```

**Review by amansinghoriginal — CHANGES_REQUESTED** [review]

File: core/src/evaluation/functions/numeric/exp.rs (hunk 4 above)
Comment (kaffie-1517) [reply core/src/evaluation/functions/numeric/exp.rs]: Agreed. I've updated the code to remove the redundant is_infinite() and is_nan() checks. Thanks for pointing that out.

File: core/src/evaluation/functions/numeric/exp.rs (hunk 3 above)
Comment (kaffie-1517) [reply core/src/evaluation/functions/numeric/exp.rs]: Thanks for catching that! You are right, I've removed the redundant lines to simplify the logic.

File: functions-gql/src/lib.rs [hunk 8]
```diff
@@ -65,6 +65,8 @@ fn register_text_functions(registry: &FunctionRegistry) {
 fn register_numeric_functions(registry: &FunctionRegistry) {
     registry.register_function("abs", Function::Scalar(Arc::new(Abs {})));
     registry.register_function("ceil", Function::Scalar(Arc::new(Ceil {})));
+    registry.register_function("e", Function::Scalar(Arc::new(E {})));
```
Comment (amansinghoriginal) [inline functions-gql/src/lib.rs]: Remove this line. The GQL standard (ISO/IEC 39075) defines EXP() but not a standalone e() constant. GQL users get Euler's number via EXP(1).

File: core/src/evaluation/functions/numeric/exp.rs [hunk 9]
```diff
@@ -0,0 +1,80 @@
...
+            });
+        }
+        match &args[0] {
+            VariableValue::Null => Ok(VariableValue::Null),
+            VariableValue::Integer(n) => match n.as_i64() {
+                Some(i) => {
+                    let result = (i as f64).exp();
+                    match Float::from_f64(result) {
+                        Some(f) => Ok(VariableValue::Float(f)),
+                        None => Err(FunctionError {
```
Comment (amansinghoriginal) [inline core/src/evaluation/functions/numeric/exp.rs]: Nit: The existing convention in sin.rs/cos.rs/tan.rs separates computation from the Float::from_f64 check using a let value = ... intermediate. Consider matching that pattern for consistency:
```rust
  let value = match n.as_i64() {
      Some(i) => (i as f64).exp(),
      None => { return Err(...) }
  };
  Ok(VariableValue::Float(match Float::from_f64(value) {
      Some(f) => f,
      None => { return Err(...) }
  }))
```

File: core/src/evaluation/functions/numeric/tests/e_tests.rs [hunk 10]
```diff
@@ -0,0 +1,71 @@
...
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::sync::Arc;
+
+use crate::evaluation::context::QueryVariables;
+use crate::evaluation::functions::numeric::E;
+use crate::evaluation::functions::ScalarFunction;
+use crate::evaluation::variable_value::float::Float;
```
Comment (amansinghoriginal) [inline core/src/evaluation/functions/numeric/tests/e_tests.rs]: Unused import? Is `Float` used in this file?

File: core/src/evaluation/functions/numeric/tests/e_tests.rs [hunk 11]
```diff
@@ -0,0 +1,71 @@
...
+    // Test with 1 argument (should be 0)
+    let result = func
+        .call(
+            &context,
+            &expression,
+            vec![VariableValue::Integer(1.into())],
+        )
+        .await;
+
+    use crate::evaluation::{FunctionError, FunctionEvaluationError};
```
Comment (amansinghoriginal) [inline core/src/evaluation/functions/numeric/tests/e_tests.rs]: Can we move this to the module-level imports at top of the file, to stay consistent with the codebase.

File: core/src/evaluation/functions/numeric/tests/exp_tests.rs [hunk 12]
```diff
@@ -0,0 +1,183 @@
...
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::sync::Arc;
+
+use crate::evaluation::context::QueryVariables;
+use crate::evaluation::functions::numeric::Exp;
+use crate::evaluation::functions::ScalarFunction;
+use crate::evaluation::variable_value::float::Float;
```
Comment (amansinghoriginal) [inline core/src/evaluation/functions/numeric/tests/exp_tests.rs]: Unused import? Float is used in assertions but is only needed for constructing expected values. Other test files (e.g., abs_tests.rs) use assert_eq!(result, 1.0_f64) directly without

File: core/src/evaluation/functions/numeric/tests/exp_tests.rs [hunk 13]
```diff
@@ -0,0 +1,183 @@
...
+use crate::evaluation::variable_value::VariableValue;
+use crate::evaluation::{ExpressionEvaluationContext, FunctionEvaluationError, InstantQueryClock};
+use drasi_query_ast::ast;
+
+#[tokio::test]
+async fn test_exp() {
```
Comment (amansinghoriginal) [inline core/src/evaluation/functions/numeric/tests/exp_tests.rs]: This bundles 4 scenarios (integer 0, integer 1, float 2.0, negative -1) into one test. The codebase convention (abs_tests.rs, sin_tests.rs) uses one test per scenario. Like `exp_zero`, `exp_float`, `exp_integer`, `exp_negative`. This gives clearer failure identification and matches the existing pattern.

File: core/src/evaluation/functions/numeric/tests/exp_tests.rs [hunk 14]
```diff
@@ -0,0 +1,183 @@
...
+        name: Arc::from("exp"),
+        args: vec![],
+        position_in_query: 0,
+    };
+
+    // Test with 0 arguments (should be 1)
+    let result = func
+        .call(&context, &expression, vec![])
+        .await;
+    use crate::evaluation::{FunctionError, FunctionEvaluationError};
```
Comment (amansinghoriginal) [inline core/src/evaluation/functions/numeric/tests/exp_tests.rs]: move to module-level imports. FunctionEvaluationError is already imported at line 23, so this might creates a duplicate.

File: core/src/evaluation/functions/numeric/tests/exp_tests.rs [hunk 15]
```diff
@@ -0,0 +1,183 @@
...
+        .call(
+            &context,
+            &expression,
+            vec![VariableValue::Float(Float::from_f64(2.0).unwrap())],
+        )
+        .await
+        .unwrap();
+    match result {
+        VariableValue::Float(f) => {
+            assert!((f.as_f64().unwrap() - std::f64::consts::E.powf(2.0)).abs() < 1e-10);
```
Comment (amansinghoriginal) [inline core/src/evaluation/functions/numeric/tests/exp_tests.rs]: f64::exp() is deterministic... the same input always produces the same f64 result. Exact `assert_eq!` works here and is what the rest of the codebase uses (see pi_tests.rs, abs_tests.rs).

Approximate comparison is unnecessary and masks potential issues.

File: core/src/evaluation/functions/numeric/tests/exp_tests.rs [hunk 16]
```diff
@@ -0,0 +1,183 @@
...
+            &context,
+            &expression,
+            vec![VariableValue::Integer((-1).into())],
+        )
+        .await
+        .unwrap();
+    match result {
+        VariableValue::Float(f) => {
+            // exp(-1) = 1/e
+            assert!((f.as_f64().unwrap() - (-1.0f64).exp()).abs() < 1e-10);
```
Comment (amansinghoriginal) [inline core/src/evaluation/functions/numeric/tests/exp_tests.rs]: Same issue - can we use exact deterministic f64::exp() here?

**Review by amansinghoriginal — CHANGES_REQUESTED** [review]
Thanks for the PR and updates so far.
Please resolve the newly identified issues.

File: functions-gql/src/lib.rs (hunk 8 above)
Comment (amansinghoriginal) [reply functions-gql/src/lib.rs]: @kaffie-1517 Please take a look at this.

Comment (kaffie-1517) [reply functions-gql/src/lib.rs]: Hey, @amansinghoriginal, working on it.
